modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -702,46 +702,50 @@ private CompletableFuture<Void> beforeZoneReplicaStartedImpl(LocalPartitionRepli
-                                                        lwm
-                                                )
-                                        );
-
-                                        preparePartitionResourcesAndLoadToZoneReplica(tbl, zonePartitionId, parameters.onRecovery());
-                                    }), ioExecutor)
-                                    // If the table is already closed, it's not a problem (probably the node is stopping).
-                                    .exceptionally(ignoreTableClosedException());
-                        }))
-                        .toArray(CompletableFuture[]::new);
+                return readLockAcquisitionFuture.thenCompose(stamp -> {
+                    Set<TableImpl> zoneTables = zoneTablesRawSet(zonePartitionId.zoneId());
+
+                    int partitionIndex = zonePartitionId.partitionId();
+
+                    PartitionSet singlePartitionIdSet = PartitionSet.of(partitionIndex);
+
+                    CompletableFuture<?>[] futures = zoneTables.stream()
+                            .map(tbl -> inBusyLockAsync(busyLock, () -> {
+                                return getOrCreatePartitionStorages(tbl, singlePartitionIdSet)
+                                        .thenRunAsync(() -> inBusyLock(busyLock, () -> {
+                                            localPartsByTableId.compute(
+                                                    tbl.tableId(),
+                                                    (tableId, oldPartitionSet) -> extendPartitionSet(oldPartitionSet, partitionIndex)
+                                            );
+
+                                            lowWatermark.getLowWatermarkSafe(lwm ->
+                                                    registerIndexesToTable(
+                                                            tbl,
+                                                            catalogService,
+                                                            singlePartitionIdSet,
+                                                            tbl.schemaView(),
+                                                            lwm
+                                                    )
+                                            );
+
+                                            preparePartitionResourcesAndLoadToZoneReplica(tbl, zonePartitionId, parameters.onRecovery());
+                                        }), ioExecutor)
+                                        // If the table is already closed, it's not a problem (probably the node is stopping).
+                                        .exceptionally(ignoreTableClosedException());
+                            }))
+                            .toArray(CompletableFuture[]::new);
 
-                return allOf(futures).thenAccept(unused -> zoneLock.unlockRead(stamp));
+                    return allOf(futures).thenAccept(unused -> zoneLock.unlockRead(stamp));
```

Comment:
If any of `futures` completes exceptionally, `thenAccept()` will not be called, so the lock will remain forever. Should we switch to `whenComplete()`?

Reply:
Nice catch! Fixed.